File: 2021/day_15/main.go
package main

import (
	"adventofcode/adventUtils"
	"fmt"
	"math"
)

type coord [2]int
type cave map[coord]int

var DIRECTIONS = [4]coord{coord{1, 0}, coord{0, 1}, coord{-1, 0}, coord{0, -1}}

func parseLines(l []string) cave {
	coordMap := cave{}
	for i, row := range l {
		for j, value := range []rune(row) {
			v := int(value) - '0'
			coordMap[coord{i, j}] = v
		}
	}
	return coordMap
}

func getMinGuess(coords []coord, guessMap map[coord]int) (coord, []coord) {
	var min coord
	minIndex := 0
	minCost := math.MaxInt32
	for i, c := range coords {
		if v, _ := guessMap[c]; v < minCost {
			min = c
			minCost = v
			minIndex = i
		}
	}
	coords[minIndex] = coords[len(coords)-1]
	return min, coords[:len(coords)-1]
}

func isInCoord(i coord, l []coord) bool {
	for _, j := range l {
		if i == j {
			return true
		}
	}
	return false
}

func findLowRiskRoute(c cave) int {
	// Stolen from https://en.wikipedia.org/wiki/A*_search_algorithm
	score := map[coord]int{}
	guess := map[coord]int{}

	maxX, maxY := 0, 0

	for k := range c {
		score[k] = math.MaxInt32
		guess[k] = 0
		if k[0] > maxX {
			maxX = k[0]
		}
		if k[1] > maxY {
			maxY = k[1]
		}
	}

	start := coord{0, 0}
	goal := coord{maxX, maxY}

	toCheck := []coord{start}
	current := start

	score[start] = 0
	guess[start] = maxX + maxY

	for len(toCheck) > 0 {
		current, toCheck = getMinGuess(toCheck, guess)
		if current == goal {
			s, ok := score[current]
			if !ok {
				panic("Something went wrong")
			}
			return s
		}

		currCost, _ := score[current]

		for _, d := range DIRECTIONS {
			next := coord{current[0] + d[0], current[1] + d[1]}
			enterCost, ok := c[next]
			if !ok {
				continue
			}

			nextCost, _ := score[next]
			tentativeCost := currCost + enterCost
			if tentativeCost > nextCost {
				continue
			}
			score[next] = tentativeCost
			guess[next] = tentativeCost + (maxX - next[0]) + (maxY - next[1])
			if !isInCoord(next, toCheck) {
				toCheck = append(toCheck, next)
			}
		}
	}

	panic("Could not reach goal")
}

func findBigLowRiskRoute(c cave) int {
	newCave := cave{}
	maxX, maxY := 0, 0
	for k := range c {
		if k[0]+1 > maxX {
			maxX = k[0] + 1
		}
		if k[1]+1 > maxY {
			maxY = k[1] + 1
		}
	}
	for i := 0; i < 5; i++ {
		for j := 0; j < 5; j++ {
			for k, v := range c {
				next := coord{k[0] + maxX*i, k[1] + maxX*j}
				newCave[next] = (v-1+i+j)%9 + 1
			}
		}
	}
	return findLowRiskRoute(newCave)
}

func day15() {
	filename := "input.txt"
	rawLines, err := adventUtils.ReadFileAdvent(filename, 15)
	if err != nil {
		fmt.Errorf("%+v\n", err)
		panic(err)
	}
	c := parseLines(rawLines)

	// Part 1
	risk := findLowRiskRoute(c)
	fmt.Printf("The total riskiness is %d\n", risk)

	// Part 2
	bigRisk := findBigLowRiskRoute(c)
	fmt.Printf("The total riskiness of the big map is %d\n", bigRisk)

}

func main() {
	adventUtils.Benchmark(day15)
}

File: 2021/day_15/main.rs
use std::cmp;
use std::collections::binary_heap::BinaryHeap;
use std::collections::{HashMap, HashSet};
use std::fs;
use std::time::Instant;

const DIRECTIONS: [(i32, i32); 4] = [(1, 0), (0, 1), (-1, 0), (0, -1)];

#[derive(cmp::PartialEq, cmp::Eq, cmp::PartialOrd, cmp::Ord)]
struct HeapPosition {
    estimate: i32,
    coord: (i32, i32),
}

fn parse_file(path: &str) -> HashMap<(i32, i32), i32> {
    let parsed = fs::read_to_string(path).expect("Could not read file");
    let raw_cave = parsed.lines().into_iter().map(|line| {
        line.chars().into_iter().map(|c| {
            c.to_string()
                .parse::<i32>()
                .expect("Could not parse character to number")
        })
    });
    let mut cave = HashMap::<(i32, i32), i32>::new();
    for (i, l) in raw_cave.enumerate() {
        for (j, e) in l.enumerate() {
            cave.insert((i as i32, j as i32), e);
        }
    }
    cave
}

fn a_star_search(cave: &HashMap<(i32, i32), i32>) -> i32 {
    let mut score = HashMap::<(i32, i32), i32>::new();
    let mut guess = HashMap::<(i32, i32), i32>::new();

    let (mut max_x, mut max_y) = (0, 0);
    for (k, _) in cave {
        score.insert(*k, i32::MAX);
        guess.insert(*k, 0);
        max_x = cmp::max(max_x, k.0);
        max_y = cmp::max(max_y, k.1);
    }

    let mut to_check = BinaryHeap::new();
    let mut to_check_set = HashSet::new();
    to_check.push(cmp::Reverse(HeapPosition {
        estimate: max_x + max_y,
        coord: (0, 0),
    }));
    to_check_set.insert((0, 0));
    score.insert((0, 0), 0);
    let goal = (max_x, max_y);

    while !to_check.is_empty() {
        let current = to_check.pop().expect("Could not find a new item").0;
        if current.coord == goal {
            return score[&current.coord];
        }
        to_check_set.remove(&current.coord);

        let current_cost = score[&current.coord];

        for d in DIRECTIONS {
            let next = (current.coord.0 + d.0, current.coord.1 + d.1);
            let enter_cost: i32;
            match cave.get(&next) {
                Some(v) => enter_cost = *v,
                None => continue,
            }

            let next_cost = score[&next];
            let tentative = current_cost + enter_cost;
            if tentative > next_cost {
                continue;
            }
            score.insert(next, tentative);
            guess.insert(next, tentative + (max_x - next.0) + (max_y - next.1));
            if !to_check_set.contains(&next) {
                to_check_set.insert(next);
                to_check.push(cmp::Reverse(HeapPosition {
                    coord: next,
                    estimate: tentative,
                }))
            }
        }
    }
    panic!("Could not find the path")
}

fn big_risk_search(cave: &HashMap<(i32, i32), i32>) -> i32 {
    let mut big_cave = HashMap::new();

    let (mut max_x, mut max_y) = (0, 0);
    for (k, _) in cave {
        max_x = cmp::max(max_x, k.0);
        max_y = cmp::max(max_y, k.1);
    }

    for i in 0..5 {
        for j in 0..5 {
            for (k, v) in cave {
                let next = (k.0 + max_x * i, k.1 + max_y * j);
                big_cave.insert(next, (v - 1 + i + j) % 9 + 1);
            }
        }
    }
    a_star_search(&big_cave)
}

fn day_15() {
    let filename = "2021/day_15/input.txt";
    let cave = parse_file(filename);

    // Part 1
    let risk = a_star_search(&cave);
    println!("The total risk is {}", risk);

    // Part 2
    let big_risk = big_risk_search(&cave);
    println!("The total risk for the big champer is {}", big_risk);
}

fn main() {
    let now = Instant::now();
    day_15();
    println!("Rust took {} microseconds", now.elapsed().as_micros());
}
